File: examples/python/diagnostic_tools.py
#! /usr/bin/python
# Very basic script demonstrating diagnostic tools functionality
#
""" Copyright 2015 Akamai Technologies, Inc. All Rights Reserved.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.

 You may obtain a copy of the License at 

    http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
"""
import requests, logging, json, sys
from random import randint
from akamai.edgegrid import EdgeGridAuth
from config import EdgeGridConfig
from urlparse import urljoin
import urllib
session = requests.Session()
debug = False
section_name = "default"

# If all parameters are set already, use them.  Otherwise
# use the config
try:
	config = EdgeGridConfig({"verbose":debug},section_name)
except:
	error_msg = "ERROR: No section named %s was found in your ~/.edgerc file\n" % section_name
	error_msg += "ERROR: Please generate credentials for the script functionality\n"
	error_msg += "ERROR: and run 'gen_edgerc %s' to generate the credential file\n" % section_name
	sys.exit(error_msg)

if config.verbose or config.debug:
  debug = True


# Set the config options
session.auth = EdgeGridAuth(
            client_token=config.client_token,
            client_secret=config.client_secret,
            access_token=config.access_token
)

# Set the baseurl based on config.host
baseurl = '%s://%s/' % ('https', config.host)

def getResult(endpoint, parameters=None):
  if parameters:
    parameter_string = urllib.urlencode(parameters)
    path = ''.join([endpoint + '?',parameter_string])
  else:
    path = endpoint
  endpoint_result = session.get(urljoin(baseurl,path))
  httpErrors(endpoint_result.status_code, path, endpoint_result.json())
  if debug: print ">>>\n" + json.dumps(endpoint_result.json(), indent=2) + "\n<<<\n"
  return endpoint_result.json()

def httpErrors(status_code, endpoint, result):
  if status_code == 403:
                error_msg =  "ERROR: Call to %s failed with a 403 result\n" % endpoint
                error_msg +=  "ERROR: This indicates a problem with authorization.\n"
                error_msg +=  "ERROR: Please ensure that the credentials you created for this script\n"
                error_msg +=  "ERROR: have the necessary permissions in the Luna portal.\n"
                error_msg +=  "ERROR: Problem details: %s\n" % result["detail"]
                exit(error_msg)

  if status_code in [400, 401]:
                error_msg =  "ERROR: Call to %s failed with a %s result\n" % (endpoint, status_code)
                error_msg +=  "ERROR: This indicates a problem with authentication or headers.\n"
                error_msg +=  "ERROR: Please ensure that the .edgerc file is formatted correctly.\n"
                error_msg +=  "ERROR: If you still have issues, please use gen_edgerc.py to generate the credentials\n"
                error_msg +=  "ERROR: Problem details: %s\n" % result["detail"]
                exit(error_msg)

  if status_code in [404]:
                error_msg =  "ERROR: Call to %s failed with a %s result\n" % (endpoint, status_code)
                error_msg +=  "ERROR: This means that the page does not exist as requested.\n"
                error_msg +=  "ERROR: Please ensure that the URL you're calling is correctly formatted\n"
                error_msg +=  "ERROR: or look at other examples to make sure yours matches.\n"
                error_msg +=  "ERROR: Problem details: %s\n" % result["detail"]
                exit(error_msg)

  error_string = None
  if "errorString" in result:
               if result["errorString"]:
                       error_string = result["errorString"]
  else:
    for key in result:
      if type(key) is not str:
        continue
      if type(result[key]["errorString"]) is str:
        error_string = result[key]["errorString"]
  if error_string:
                error_msg =  "ERROR: Call caused a server fault.\n"
                error_msg +=  "ERROR: Please check the problem details for more information:\n"
                error_msg +=  "ERROR: Problem details: %s\n" % error_string
                exit(error_msg) 
		

# Request locations that support the diagnostic-tools
print
print "Requesting locations that support the diagnostic-tools API.\n"

location_result = getResult('/diagnostic-tools/v1/locations')

# Select a random location to host our request
location_count = len(location_result['locations'])

print "There are %s locations that can run dig in the Akamai Network" % location_count
rand_location = randint(0, location_count-1)
location = location_result['locations'][rand_location]
print "We will make our call from " + location + "\n"

# Request the dig request the {OPEN} Developer Site IP informantion
print "Running dig from " + location
dig_parameters = { "hostname":"developer.akamai.com", "location":location, "queryType":"A" }
dig_result = getResult("/diagnostic-tools/v1/dig",dig_parameters)

# Display the results from dig
print dig_result['dig']['result']

File: examples/node/src/diagnostic-tools.js
#! /usr/bin/env node

/**
Copyright 2015 Akamai Technologies, Inc. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.

You may obtain a copy of the License at 

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/

/**
 * Example of using the Akamai OPEN diagnostic-tools API
 * https://developer.akamai.com/api/luna/diagnostic-tools/reference.html
 */

var path = require('path');
var os = require('os');
var prettyJSON = require('prettyjson');
var argv = require('minimist')(process.argv.slice(2));
var EdgeGrid = require('edgegrid');

var debug = argv.debug ? true : false;
var verbose = argv.verbose ? true : false;
var sectionName = "default";
var edgercPath = path.join(os.homedir(), "/.edgerc");
var headers = argv.headers ? argv.headers : {};

var eg = new EdgeGrid({
  path: edgercPath,
  section: sectionName,
  debug: debug
});

eg.auth({
  path: '/diagnostic-tools/v1/locations',
  method: 'GET',
  headers: {},
  body: {}
});

eg.send(function(data, response) {
  console.log("\nRequesting locations that support the diagnostic-tools API.");
  console.log("Data: ", data);

  // Convert data to JSON format
  data = JSON.parse(data);

  // Pick random location
  var locationCount = data.locations.length;
  var location = data.locations[Math.floor(Math.random() * locationCount)];

  console.log("There are " + locationCount + " locations that can run dig in the Akamai Network.");
  console.log("We will make our call from " + location);

  // Prettify the output and list locations
  // console.log("\nData received from diagnostic-tools request:");
  // console.log(prettyJSON.render(data));

  // Start dig request
  makeDigRequest(location);
});

function makeDigRequest(location) {
  console.log("Starting dig request, this may take a moment...");
  // Perform dig request for developer.akamai.com using the Akamai server 
  // location specified and query type 'A' which performs a mapping to an IPv4
  // address
  var digParameters = {
    "hostname": "developer.akamai.com",
    "queryType": "A",
    "location": location
  };

  // Call auth, passing the query string parameters in via the 'qs' property
  eg.auth({
    path: '/diagnostic-tools/v1/dig',
    method: 'GET',
    headers: {},
    body: {},
    qs: digParameters
  });

  eg.send(function(data, response) {
    // Convert data to JSON format
    data = JSON.parse(data);

    // Display the results from dig
    console.log("\nData received from dig request:");
    console.log(prettyJSON.render(data));
  });
}
